torch/csrc/jit/api/module.h

```diff
@@ -141,8 +141,35 @@ struct TORCH_API Module : public Object {
     _ivalue()->setAttr(name, std::move(v));
   }
 
-  void register_module(const std::string& name, const Module& module) {
-    type()->addOrCheckAttribute(name, module.type());
+  void register_module(
+      const std::string& name,
+      const Module& module,
```

Comment:
It seems to me that this is a special case (where we want to add a module attribute but register its type as an interface). My suggestion is to avoid changing the public module API and do this in the implementation of cloning, to hide this from users. We already have special rules for module attributes there, so we can just change that rule to do:
1. Copy the attribute type
2. Insert the cloned submodule into the ivalue.

Reply:
sounds good, I will change it accordingly.

Reply:
I took a second thought on this. Right now in python, we have a way for user to define an interface attr with a submodule, but we don't have a equivalent way to do this in C++. Does it mean that right now if we do it in C++, user have to call `register_attribute` for a module interface type? that seems a bit weird to me.

Reply:
I think we don't intend for people to construct modules with interface types directly from C++ (for example, the C++ constructors for interfaces are not part of the public API in any way). If we want to add that functionality, we would need to add a lot more than just the bindings on `Module`, so my preference is to leave it out for now.